ceno_zkvm/src/instructions/riscv/arith.rs

```diff
@@ -136,184 +136,76 @@ mod test {
-                    0,
-                ),
-            ])
-            .unwrap();
-
-        let expected_rd_written = UInt::from_const_unchecked(
-            Value::new_unchecked(11_u32.wrapping_sub(2))
-                .as_u16_limbs()
-                .to_vec(),
-        );
-
-        config
-            .rd_written
-            .require_equal(|| "assert_rd_written", &mut cb, &expected_rd_written)
-            .unwrap();
-
-        MockProver::assert_satisfied_raw(&cb, raw_witin, &[insn_code], None, Some(lkm));
+        verify::<SubOp>("basic", 11, 2);
+        verify::<SubOp>("0 - 0", 0, 0);
+        verify::<SubOp>("1 - 0", 1, 0);
+        verify::<SubOp>("1 - 1", 1, 1);
+        verify::<SubOp>("underflow", 3, 11);
     }
 
-    #[test]
-    fn test_opcode_sub_underflow() {
+    fn verify<I: RIVInstruction>(name: &'static str, rs1: u32, rs2: u32) {
         let mut cs = ConstraintSystem::<GoldilocksExt2>::new(|| "riscv");
         let mut cb = CircuitBuilder::new(&mut cs);
         let config = cb
             .namespace(
-                || "sub",
-                |cb| {
-                    let config = SubInstruction::construct_circuit(cb);
-                    Ok(config)
-                },
+                || format!("{:?}_({name})", I::INST_KIND),
+                |cb| Ok(ArithInstruction::<GoldilocksExt2, I>::construct_circuit(cb)),
             )
             .unwrap()
             .unwrap();
 
-        let insn_code = encode_rv32(InsnKind::SUB, 2, 3, 4, 0);
-        let (raw_witin, _) =
-            SubInstruction::assign_instances(&config, cb.cs.num_witin as usize, vec![
-                StepRecord::new_r_instruction(
-                    3,
-                    MOCK_PC_START,
-                    insn_code,
-                    3,
-                    11,
-                    Change::new(0, 3_u32.wrapping_sub(11)),
-                    0,
-                ),
-            ])
-            .unwrap();
+        let outcome = match I::INST_KIND {
+            InsnKind::ADD => rs1.wrapping_add(rs2),
+            InsnKind::SUB => rs1.wrapping_sub(rs2),
+            _ => unreachable!("Unsupported instruction kind"),
+        };
 
-        let expected_rd_written = UInt::from_const_unchecked(
-            Value::new_unchecked(3_u32.wrapping_sub(11))
-                .as_u16_limbs()
-                .to_vec(),
-        );
+        // values assignment
+        let insn_code = encode_rv32(I::INST_KIND, 2, 3, 4, 0);
+        let (raw_witin, lkm) = ArithInstruction::<GoldilocksExt2, I>::assign_instances(
+            &config,
+            cb.cs.num_witin as usize,
+            vec![StepRecord::new_r_instruction(
+                3,
+                MOCK_PC_START,
+                insn_code,
+                rs1,
+                rs2,
+                Change::new(0, outcome),
+                0,
+            )],
+        )
+        .unwrap();
+
+        // verify value write to register, which is only hi
```

Comment:
This comment need a cosmetics

Reply:
No idea where this came from 🤣 , addressed in [b49dc23](https://github.com/scroll-tech/ceno/pull/685/commits/b49dc238a9624bb7581ba010bd6e6e2f1a35c232)